ebos/ecloutputblackoilmodule.hh

```diff
@@ -772,6 +862,106 @@ private:
+     *    dynamic information.
+     *
+     * \param[in] scvfIdx Linear index of current interior bulk connection.
+     *
+     * \param[in] timeIdx Historical time-point at which to evaluate dynamic
+     *    quantities (e.g., reciprocal FVF or dissolved gas concentration).
+     *    Zero for the current time.
+     *
+     * \return Surface level component flow rates.
+     */
+    data::InterRegFlowMap::FlowRates
+    getComponentSurfaceRates(const ElementContext& elemCtx,
+                             const Scalar          faceArea,
+                             const std::size_t     scvfIdx,
+                             const std::size_t     timeIdx) const
+    {
+        using Component = data::InterRegFlowMap::Component;
+
+        auto rates = data::InterRegFlowMap::FlowRates{};
+
+        const auto& extQuant = elemCtx.extensiveQuantities(scvfIdx, timeIdx);
+
+        const auto alpha = getValue(extQuant.extrusionFactor()) * faceArea;
+
+        if (FluidSystem::phaseIsActive(oilPhaseIdx)) {
+            const auto& up = elemCtx
+                .intensiveQuantities(extQuant.upstreamIndex(oilPhaseIdx), timeIdx);
+
+            using FluidState = std::remove_cv_t<std::remove_reference_t<
+                decltype(up.fluidState())>>;
+
+            const auto pvtReg = up.pvtRegionIndex();
+
+            const auto bO = getValue(getInvB_<FluidSystem, FluidState, Scalar>
+                                     (up.fluidState(), oilPhaseIdx, pvtReg));
+
+            const auto qO = alpha * bO * getValue(extQuant.volumeFlux(oilPhaseIdx));
+
+            rates[Component::Oil] += qO;
+
+            if (FluidSystem::phaseIsActive(gasPhaseIdx)) {
+                const auto Rs = getValue(
+                    BlackOil::getRs_<FluidSystem, FluidState, Scalar>
+                    (up.fluidState(), pvtReg));
+
+                rates[Component::Gas]    += qO * Rs;
+                rates[Component::Disgas] += qO * Rs;
+            }
+        }
+
+        if (FluidSystem::phaseIsActive(gasPhaseIdx)) {
+            const auto& up = elemCtx
+                .intensiveQuantities(extQuant.upstreamIndex(gasPhaseIdx), timeIdx);
+
+            using FluidState = std::remove_cv_t<std::remove_reference_t<
+                decltype(up.fluidState())>>;
+
+            const auto pvtReg = up.pvtRegionIndex();
+
+            const auto bG = getValue(getInvB_<FluidSystem, FluidState, Scalar>
+                                     (up.fluidState(), gasPhaseIdx, pvtReg));
+
+            const auto qG = alpha * bG * getValue(extQuant.volumeFlux(gasPhaseIdx));
+
+            rates[Component::Gas] += qG;
+
+            if (FluidSystem::phaseIsActive(oilPhaseIdx)) {
+                const auto Rv = getValue(
+                    BlackOil::getRv_<FluidSystem, FluidState, Scalar>
+                    (up.fluidState(), pvtReg));
+
+                rates[Component::Oil]    += qG * Rv;
+                rates[Component::Vapoil] += qG * Rv;
+            }
```

Comment:
Looks pretty similar to above besides Blackoil::getRv_ vs Blackoil::getRs_, maybe we should make a function for reuse that takes getRv_/getRs_ as a std::function?

Reply:
> Looks pretty similar to above besides `Blackoil::getRv_` vs `Blackoil::getRs_`, maybe we should make a function for reuse that takes `getRv_`/`getRs_` as a `std::function`?

Maybe.  While the control structure is indeed very similar please note that we update different items of the `rates[]` buffer.  This part updates `Component::Oil` and `Component::Vapoil` while the [other block](https://github.com/OPM/opm-simulators/blob/b9a1270bbb90e0f40004c5fab6cbc7a1402e3e50/ebos/ecloutputblackoilmodule.hh#L910-L917) updates `Component::Gas` and `Component::Disgas`.